nixos/containers: now uses nixpkgs.hostPlatform

maifel-maifel opened the pull request:
###### Description of changes

This commit fixes a build error related to nixpkgs.localSystem is being deprecated for use by modules and packages for reasons described in https://github.com/NixOS/nixpkgs/issues/49765 . The new interface to be used instead is nixpkgs.hostPlatform.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] cross compiled on x86_64-linux to armv7l-linux
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

yu-re-ka commented:
Generally this is a good change and is needed for containers to work properly in nixos configurations where only hostPlatform is defined but localSystem is set to null (e.g. any modern flake nixos configuration).
_But_ it needs some more changes before we can do that, because currently in non-flake configs localSystem has a default value but hostPlatform does not, so on non-flake configs that were using the localSystem default (which is `builtins.currentSystem`), this will break.
Can be reproduced with `nix build github:maifel-maifel/nixpkgs/mr-containers-hostPlatform#nixosTests.containers-nested`.

roberth commented on nixos/modules/virtualisation/nixos-containers.nix:
```diff
@@ -483,7 +483,7 @@ in
                         _file = "module at ${__curPos.file}:${toString __curPos.line}";
                         config = {
-                          nixpkgs = { inherit localSystem; };
+                          nixpkgs = { inherit hostPlatform; };
```
Maybe something like this, in order to support the old style as well.

```suggestion
                          nixpkgs =
                            if options.nixpkgs.hostPlatform.isDefined
                              then { inherit hostPlatform; }
                              else { inherit localSystem};
```

roberth commented:
I've changed it so that non-hostPlatform configurations don't see any change yet.

@ofborg test containers-bridge
@ofborg test containers-custom-pkgs
@ofborg test containers-ephemeral
@ofborg test containers-extra_veth
@ofborg test containers-hosts
@ofborg test containers-imperative
@ofborg test containers-ip
@ofborg test containers-macvlans
@ofborg test containers-names
@ofborg test containers-nested
@ofborg test containers-physical_interfaces
@ofborg test containers-portforward
@ofborg test containers-reloadable
@ofborg test containers-restart_networking
@ofborg test containers-tmpfs

yu-re-ka commented:
Do containers potentially use a different nixpkgs version for evaluation? Otherwise the `if options.nixpkgs?hostPlatform ` check would be unnecessary

roberth commented:
> @roberth do containers potentially use a different nixpkgs version for evaluating the nixos config of the container? Otherwise the `if options.nixpkgs?hostPlatform` check would be unnecessary

They do: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/nixos-containers.nix#L480

yu-re-ka commented:
Perfect, then this is reasonable. One nit: Now that we have the `@host` anyways, we can remove the `host.config.nixpkgs` instead of `hostNixpkgs`? This seems cleaner to me

Commit:
nixos/containers: support nixpkgs.hostPlatform

Use hostPlatform if both the host and the containers nixpkgs supports
hostPlatform, otherwise fall back to localSystem. This preseves backwards
compatibility.
